### PR TITLE
Codebreaker-web

### PR DESCRIPTION
**Pull request by mmrshk**
https://codebreaker-yeroshek.herokuapp.com/

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+version: 2
+jobs:
+  build:
+    docker:
+       - image: circleci/ruby:2.4.1-node-browsers
```
Comment (dmbavykin) [inline .circleci/config.yml]: why u use different versions?

File: database/data.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (dmbavykin) [inline database/data.yml]: this file should be git-ignored

File: lib/models/renderer.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+class Renderer
+  def message(phrase, hashee = {})
```
Comment (dmbavykin) [inline lib/models/renderer.rb]: do we really need it?

File: lib/models/storage.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class Storage
```
Comment (dmbavykin) [inline lib/models/storage.rb]: do we really need this class?

File: lib/racker.rb [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    Rack::Response.new(lose_view) do
+      destroy_session
+    end
+  end
+
+  def win
+    return error404_view if exist?(:game)
+
+    Rack::Response.new(win_view) do
+      game = create_game
```
Comment (dmbavykin) [inline lib/racker.rb]: do we really need to define this variable?

File: lib/racker.rb [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    @request.session[:game].code
+  end
+
+  private
+
+  def destroy_session
+    @request.session.clear
+  end
+
+  def set_quess_code
```
Comment (dmbavykin) [inline lib/racker.rb]: spelling

File: lib/racker.rb [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    return @request.session[:guess_code] unless @request.session[:guess_code].nil?
```
Comment (dmbavykin) [inline lib/racker.rb]:
```ruby
return @request.session[:guess_code] if @request.session[:guess_code]
```

File: spec/racker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      it { expect(last_response.body).to include I18n.t(:cb_rules) }
+    end
+
+    context 'when statistics path' do
+      before do
+        env 'rack.session', scores: []
+        get '/statistics'
+      end
+
+      it { expect(last_request.session[:scores]).not_to be nil }
```
Comment (dmbavykin) [inline spec/racker_spec.rb]: useless test
check scores data on the page

File: spec/racker_spec.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  describe '#hint' do
+    before do
+      game.generate(Codebreaker::Entities::Game::DIFFICULTIES[:hell])
+      env 'rack.session', game: game, used_hints: []
+      get '/hint'
+    end
+
+    it 'adds value to session hints array' do
+      post '/hint'
+      expect(game.hints).to be_empty
```
Comment (dmbavykin) [inline spec/racker_spec.rb]: u should not test gem logic

File: spec/racker_spec.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      expect(last_request.session[:used_hints]).not_to be_empty
```
Comment (6aBblKuH) [inline spec/racker_spec.rb]: Can we check presence hinted number in used_hints array?

File: lib/models/guess.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  def produce_code
+    code = start_process
+    code + FAIL_DIGIT_CHAR * (Codebreaker::Entities::Game::DIGITS_COUNT - code.size)
+  end
+
+  def valid_size?(game, guess_code)
+    start_process.size == Codebreaker::Entities::Game::DIGITS_COUNT
+  end
+
+  def start_process
```
Comment (dmbavykin) [inline lib/models/guess.rb]: u call this method few times - memorize

File: lib/racker.rb [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  def stats
+    @request.session[:scores] = @storage.load
+    stats_view
+  end
+
+  def guess
+    Rack::Response.new do |response|
+      return lose unless create_game.attempts.positive?
+
+      @guess_code = @request.params['guess_code']
```
Comment (dmbavykin) [inline lib/racker.rb]: is it need to be instance variable?

File: lib/racker.rb [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    @request.session[:used_hints]
+  end
+
+  def hint
+    Rack::Response.new do |response|
+      game = create_game
+      return error404_view if game.hints_spent?
+
+      hint = game.take_a_hint!
```
Comment (dmbavykin) [inline lib/racker.rb]: do we need this variable?

File: lib/views/game.html.erb [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+              <button type="submit" class="btn btn-primary btn-lg">
+                Submit
+              </button>
+            </form>
+            <form method="post" action="/hint">
+              <button class="btn btn-warning btn-lg float-right" role="button"
+                <% if hints_spent? %> disabled <% end %>>
+                  <%= I18n.t(:show_hint) %>
+                <% if !used_hints.empty? %>
```
Comment (dmbavykin) [inline lib/views/game.html.erb]: https://apidock.com/ruby/Enumerable/any%3F

File: spec/racker_spec.rb [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    before do
+      game.generate(Codebreaker::Entities::Game::DIFFICULTIES[:hell])
+      env 'rack.session', game: game, used_hints: [], level: :easy
+      get '/hint'
+    end
+
+    it 'adds value to session hints array' do
+      post '/hint'
+      expect(last_request.session[:used_hints]).not_to be_empty
+      expect(last_request.session[:game].code.join.include?(last_request.session[:used_hints].join)).to be true
```
Comment (dmbavykin) [inline spec/racker_spec.rb]: https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/include-matcher

File: lib/racker.rb [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+    return @request.session[:guess_code] if @request.session[:guess_code]
+
+    @request.session[:guess_code] = ''
+  end
+
+  def not_exist?(param)
+    @request.session[param].nil?
+  end
+
+  def create_game
```
Comment (dmbavykin) [inline lib/racker.rb]: try to redo with memorizing

File: lib/racker.rb [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    @request.session.clear
+  end
+
+  def set_guess_code
+    return @request.session[:guess_code] if @request.session[:guess_code]
+
+    @request.session[:guess_code] = ''
+  end
+
+  def not_exist?(param)
```
Comment (dmbavykin) [inline lib/racker.rb]: can we check for existing ?

**Review by dmbavykin — APPROVED** [review]

**Pull request merged**